File: baselibrary/src/main/java/com/jy/baselibrary/helper/LogCatHelper.java
package com.jy.baselibrary.helper;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Environment;
import android.text.TextUtils;

import com.jy.baselibrary.utils.YLogUtils;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.text.SimpleDateFormat;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.TimeUnit;

/**
 * Administrator
 * created at 2018/6/7 10:27
 * TODO:logcat写文件
 */
public class LogCatHelper {

    private static final String TAG = LogCatHelper.class.getSimpleName();
    private static LogCatHelper instance = null;

    private String dirPath;//保存路径
    //日志缓冲队列
    private ArrayBlockingQueue<String> mCacheLog = new ArrayBlockingQueue<>(500);
    //缓存区大小100k
    private final int BUFF_SIZE = 100;
    private LogThread logThread;

    public static LogCatHelper getInstance() {
        if (instance == null) {
            instance = new LogCatHelper();
        }
        return instance;
    }

    public void init(Context mContext, String path) {
        if (TextUtils.isEmpty(path)) {
            dirPath = Environment.getExternalStorageDirectory().getAbsolutePath()
                    + File.separator + "seeker" + File.separator + mContext.getPackageName();
        } else {
            dirPath = path;
        }
        File dir = new File(dirPath);
        if (!dir.exists()) {
            dir.mkdirs();
        }
    }

    /**
     * 重置
     *
     * @param mContext
     * @param path
     */
    public void reset(Context mContext, String path) {
        stop();
        init(mContext, path);
        start();
    }

    /**
     * 启动log日志保存
     */
    public void start() {
        if (TextUtils.isEmpty(dirPath)) {
            YLogUtils.INSTANCE.eTag(TAG, "请先在全局Application中调用init(mContext,path)初始化！");
            return;
        }
        logThread = new LogThread(dirPath);
        logThread.start();
    }

    public void stop() {
        if (logThread != null) {
            logThread.stopWrite();
        }
    }

    /**
     * 将日志存入缓存区
     *
     * @param msg
     */
    public void writeLogFile(String msg) {
        try {
            mCacheLog.offer(msg, 100, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private class LogThread extends Thread {

        private FileOutputStream fos;
        String dirPath;
        String fileName;
        private boolean openWrite = true;

        public LogThread(String dirPath) {
            this.dirPath = dirPath;
            try {
                fileName = FormatDate.getFormatDate();
                File file = new File(dirPath, fileName + ".log");
                if (!file.exists()) {
                    file.createNewFile();
                }
                fos = new FileOutputStream(file, true);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        @Override
        public void run() {
            try {
                filter(dirPath, fileName);
                BufferedOutputStream mOutputStream = new BufferedOutputStream(fos, BUFF_SIZE);

                while (openWrite && !isInterrupted()) {
                    if (mCacheLog.size() > 0) {
                        String msg = FormatDate.getFormatTime() + mCacheLog.poll() + "\r\n";
                        byte[] bytes = msg.getBytes();
                        mOutputStream.write(bytes, 0, bytes.length);
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                try {
                    if (fos != null) {
                        fos.close();
                        fos = null;
                    }
                } catch (Exception e2) {
                    e2.printStackTrace();
                }
            }
        }

        public void stopWrite() {
            openWrite = false;
            this.interrupt();
        }
    }

    private static void filter(String dirPath, String nowFileName) {
        try {
            int nowTime = 0;
            if (!TextUtils.isEmpty(nowFileName)) {
                nowTime = Integer.parseInt(nowFileName);
                YLogUtils.INSTANCE.iTag(TAG, "创建log文件", nowFileName, "当前时间", nowTime);
            }
            File root = new File(dirPath);
            File[] fileList = root.listFiles();
            if (fileList == null || fileList.length < 1) {
                return;
            }
            YLogUtils.INSTANCE.iTag(TAG, "过滤旧log文件，只保留3天log");
            for (File file : fileList) {
                String name = file.getName();
                if (!TextUtils.isEmpty(name) && name.endsWith(".log")) {
                    int time = Integer.parseInt(name.substring(0, name.length() - 4));
                    if (nowTime - time > 2) {
                        boolean success = file.delete();
                        YLogUtils.INSTANCE.iTag(TAG, "删除文件", name, "删除成功", success);
                    } else {
                        YLogUtils.INSTANCE.iTag(TAG, "保留文件", name);
                    }
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
            YLogUtils.INSTANCE.eTag(TAG, "过滤log失败", e.getMessage());
        }
    }

    /**
     * 获取logfile
     *
     * @return
     */
    public Map<String, File> getLogFile() {
        File root = new File(dirPath);
        File[] fileList = root.listFiles();
        Map<String, File> fileMap = new HashMap<String, File>();
        if (fileList == null) {
            return fileMap;
        }
        for (int i = 0; i < fileList.length; i++) {
            fileMap.put(fileList[i].getName(), fileList[i]);
            YLogUtils.INSTANCE.iTag(TAG, "getLogFile", fileList[i].getName());
        }
        return fileMap;
    }

    public String getDirPath() {
        return dirPath;
    }

    @SuppressLint("SimpleDateFormat")
    private static class FormatDate {

        public static String getFormatDate() {
            SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd");
            return sdf.format(System.currentTimeMillis());
        }

        public static String getFormatTime() {
            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS  ");
            return sdf.format(System.currentTimeMillis());
        }
    }
}

File: baselibrary/src/main/java/com/jy/baselibrary/helper/LogCatHelper.kt
package com.jy.baselibrary.helper

import android.content.Context
import android.os.Environment
import android.text.TextUtils
import com.jy.baselibrary.utils.YLogUtils
import com.jy.baselibrary.utils.YLogUtils.eTag
import com.jy.baselibrary.utils.YLogUtils.iTag
import java.io.BufferedReader
import java.io.File
import java.io.FileOutputStream
import java.io.InputStreamReader
import java.text.SimpleDateFormat
import java.util.*

/**
 * Administrator
 * created at 2019/10/18 17:00
 * TODO:log收集
 * 注：若path为空则保存到sdcard（需要访问权限）
 */
class LogCatHelper private constructor(mContext: Context, path: String) {
    private var dirPath: String? = null//保存路径
    private val appId: Int//应用pid
    private var logThread: Thread? = null
    /**
     * 启动log日志保存
     */
    fun start() {
        if (!YLogUtils.SHOW_LOG) return
        if (logThread == null) {
            logThread = Thread(LogRunnable(appId, dirPath))
        }
        logThread?.start()
    }

    private class LogRunnable internal constructor(pid: Int, dirPath: String?) :
        Runnable {
        private var mProcess: Process? = null
        private var fos: FileOutputStream? = null
        private var mReader: BufferedReader? = null
        private val cmd: String
        private val mPid: String
        var dirPath: String?
        var fileName: String? = null
        override fun run() {
            try {
                filter(dirPath, fileName)
                mProcess = Runtime.getRuntime().exec(cmd)
                mReader = BufferedReader(
                    InputStreamReader(mProcess?.inputStream),
                    1024
                )
                var line: String
                while (mReader!!.readLine().also { line = it } != null) {
                    if (line.isEmpty()) {
                        continue
                    }
                    if (fos != null && line.contains(mPid)) {
                        fos!!.write((FormatDate.formatTime + " " + line + "\r\n").toByteArray())
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            } finally {
                mProcess?.destroy()
                mProcess = null
                try {
                    mReader?.close()
                    mReader = null
                    fos?.close()
                    fos = null
                } catch (e2: Exception) {
                    e2.printStackTrace()
                }
            }
        }

        init {
            mPid = "" + pid
            this.dirPath = dirPath
            try {
                fileName = FormatDate.formatDate
                val file = File(dirPath, "$fileName.log")
                if (!file.exists()) {
                    file.createNewFile()
                }
                fos = FileOutputStream(file, true)
            } catch (e: Exception) {
                e.printStackTrace()
            }
            cmd = "logcat *:v | grep \"($mPid)\""
        }
    }

    private object FormatDate {
        val formatDate: String
            get() {
                val sdf = SimpleDateFormat("yyyyMMddHH", Locale.CHINA)
                return sdf.format(System.currentTimeMillis())
            }

        val formatTime: String
            get() {
                val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.CHINA)
                return sdf.format(System.currentTimeMillis())
            }
    }

    companion object {
        private var instance: LogCatHelper? = null
        private val TAG = LogCatHelper::class.java.simpleName
        /**
         * @param path log日志保存根目录
         */
        fun getInstance(mContext: Context, path: String): LogCatHelper? {
            if (instance == null) {
                instance = LogCatHelper(mContext, path)
            }
            return instance
        }

        private fun filter(dirPath: String?, nowFileName: String?) {
            try {
                var nowTime = 0
                if (null != nowFileName && nowFileName.isNotEmpty()) {
                    nowTime = nowFileName.substring(0, nowFileName.length - 2).toInt()
                    iTag(
                        TAG,
                        "创建log文件",
                        nowFileName,
                        "当前时间",
                        nowTime
                    )
                }
                val root = File(dirPath)
                val fileList = root.listFiles()
                if (fileList == null || fileList.isEmpty()) {
                    return
                }
                iTag(TAG, "过滤旧log文件，只保留当天log")
                for (file in fileList) {
                    val name = file.name
                    if (name.isNotEmpty() && name.endsWith(".log")) {
                        val time = name.substring(0, name.length - 6).toInt()
                        if (nowTime - time > 0) {
                            val success = file.delete()
                            iTag(TAG, "删除文件", name, "删除成功", success)
                        } else {
                            iTag(TAG, "保留文件", name)
                        }
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
                eTag(TAG, "过滤log失败", e.message)
            }
        }
    }

    init {
        appId = android.os.Process.myPid()
        dirPath = if (TextUtils.isEmpty(path)) {
            (Environment.getExternalStorageDirectory().absolutePath + File.separator + "seeker" + File.separator + mContext.packageName)
        } else {
            path
        }
        val dir = File(dirPath)
        if (!dir.exists()) {
            dir.mkdirs()
        }
    }
}